src/elisp/treemacs-visuals.el

```diff
@@ -240,7 +247,12 @@ inserted into `treemacs-icons-hash'."
           treemacs-icon-tag-leaf-text        (cdr (assq 'treemacs-icon-tag-leaf-text        treemacs--defaults-icons))))
   (treemacs--create-icons)
   (when (fboundp 'clear-image-cache)
-    (clear-image-cache)))
+    (clear-image-cache t))
+  (dolist (buf (buffer-list))
+    (with-current-buffer buf
+      (when (eq 'treemacs-mode major-mode)
+        (treemacs--check-window-system)
```

Comment:
That call isn't needed here. That check only needs to be made when a treemacs buffer is first created.

Reply:
That function makes some icon variables buffer-local, basically keeping (some of) the old icons bound to the buffer.

The reason I call it again is to reset those buffer-local variables, to force the buffer to adopt the new icons.

Without this call, existing Treemacs buffers will retain (most of) their old icons even after refreshing the buffer.

I know where you are coming from, calling this again does feel awkward to me.

Architecturally, I question making the variables buffer-local. They are set in `(treemacs--create-icons)` with `setq`, which means this function sets either their local or global value depending on which buffer is active when the function gets called (it probably should use `setq-default`). That also applies to `treemacs-reset-icons` since it calls `treemacs--create-icons`.

I didn't think this through, but maybe testing for `(display-graphic-p)` at the time of rendering the icons would be logic that is easier to reason with, though it would likely be less performant.

Reply:
With icon creation you've gotten involved with pretty much the ugliest part of the code base.
 * The declaration and definition order is awkward since I need to take into account `treemacs--image-creation-impossible`, but the icons should also be defined at the top level so the compiler knows those variables when requiring that module.
 * `check-window-system` is mostly a remnant of the time before treemacs buffers were frame-scoped, I had to call it many more times than I do now.
 * Being able to restore icons adds another layer on complexity.
 * Some variables also do need to be buffer-local so treemacs can correctly switch them when it detects a buffer is running in a terminal. From I you say I need to pay closer attention to which variables are global/local when making those changes, too.
 * Shifting work to render is not something I'll do unless I have no choice, and so far the system works, even if it's not the cleanest.

And now we want to complicate things some by adding lazy initialization, though I suppose it's a good opportunity to do some house keeping. Now this is quite the hornet's nest and well beyond the scope of your PR, so it's all up to you if you want to go along with all those refactoring plans. I'm fine with merging this with the few changes that need to be made now and cleaning things up later.

Reply:
Thank you for the insights into the evolution of the code, I think I understand the level of complexity involved better now.

> And now we want to complicate things some by adding lazy initialization, though I suppose it's a good opportunity to do some house keeping. Now this is quite the hornet's nest and well beyond the scope of your PR, so it's all up to you if you want to go along with all those refactoring plans. I'm fine with merging this with the few changes that need to be made now and cleaning things up later.

Well, my aim is not to complicate, quite the opposite. Let's take a step back. This pull request is just meant to make it possible to resize the icons, so let's focus on that. Whether to implement lazy initialization or do house-keeping is something that can be tackled later.

The pull request as-is fulfills the original purpose of making it possible to change the icon size, just with one caveate: if `treemacs` has been loaded already, it is necessary to not only change the value of `treemacs-resize-icons`, but also to call `(treemacs-reset-icons)` to re-load the icon cache. Most users will customize settings in `use-package`'s `:config` section, so `treemacs` will have been loaded already.

My suggestion is to accept this caveate for now, because accepting this caveate is much easier than a re-work of the icon initialization logic. I really don't think I have sufficient time available to commit to that.

I could just document the caveate, in the `defcustom treemacs-resize-icons` help text and the `README`.

Consequently calling `treemacs-reset-icons` would also be have to worked into the usage example in the `README`, since that one shows the variable that can be customized. But I imagine would be a bit awkward to explain why/when you need to call it: only if `treemacs-reset-icons` is not `nil`.

Now that I think about it, to make the usage example less awkward, in my opinion a viable option would be to replace the `defcustom treemacs-resize-icons` with a function of the same name that does all the necessary steps: updates a package-private `defvar` to remember the new icon size, and calls `treemacs-reset-icons` afterwards. Then all that would be needed in regards for documentation is to document the existence of the function, and show how it can be called in the usage example. It makes the documentation simpler because there is no remind users to call `treemacs-reset-icons` themselves, nor can users forget to call it.

Reply:
> If you are ok with this as a solution, I can merge the changes from that branch into this pull request.

I think an interactive function is an even better solution for this use-case than a def-custom since a function allows for interactive changes on the fly. It also opens up the option to later create functions that will interactively scale the icons by single pixels.